src/AgentImplementation.sol

```diff
@@ -55,40 +54,65 @@ contract AgentImplementation is IAgent, ERC721Holder, ERC1155Holder {
 
     /// @notice Initialize user's agent and can only be called once.
     function initialize() external {
-        if (_caller != address(0)) revert Initialized();
-        _caller = router;
+        if (_callbackWithCharge != bytes32(0)) revert Initialized();
+        _callbackWithCharge = _INIT_CALLBACK_WITH_CHARGE;
     }
 
-    /// @notice Execute arbitrary logics
+    /// @notice Execute arbitrary logics and is only callable by the router. Charge fee based on the scenarios defined
+    ///         in the router.
+    /// @dev The router is designed to prevent reentrancy so additional prevention is not needed here
+    /// @param logics Array of logics to be executed
+    /// @param tokensReturn Array of ERC-20 tokens to be returned to the current user
+    function execute(IParam.Logic[] calldata logics, address[] calldata tokensReturn) external payable {
+        if (msg.sender != router) revert NotRouter();
+
+        _chargeFeeByMsgValue();
+
+        _executeLogics(logics, true);
+
+        _returnTokens(tokensReturn);
+    }
+
+    /// @notice Execute arbitrary logics and is only callable by the router using a signer's signature
+    /// @dev The router is designed to prevent reentrancy so additional prevention is not needed here
     /// @param logics Array of logics to be executed
     /// @param fees Array of fees
     /// @param tokensReturn Array of ERC-20 tokens to be returned to the current user
-    function execute(
+    function executeWithSignature(
         IParam.Logic[] calldata logics,
         IParam.Fee[] calldata fees,
         address[] calldata tokensReturn
-    ) external payable checkCaller {
-        _executeLogics(logics);
+    ) external payable {
+        if (msg.sender != router) revert NotRouter();
 
```

Comment:
Is it possible to consider transforming it into a modifier? Becuase it is used in `execute` and `executeWithSignature` functions.

Reply:
Didn't do that since `executeByCallback` checks `msg.sender` inside the function, and would like to align the style of 3 `execute*` functions. I'm open to adding the `onlyRouter` modifier if that's your preference.

Reply:
I understand now. It's acceptable to me if you prefer to align the style of the three execute() functions.